Split CoNLL lines at single tabs

MrEbbinghaus opened the pull request:
CoNLL should be separated by a single tab.
Some CoNLL files (like the Hamburg Dependency Treebank) are treating words like "New York" as a single form with a space!
The CoNLLReader couldn't read those files.

CoNLL-X Specification defines TAB as the separator:
http://anthology.aclweb.org/W/W06/W06-2920.pdf
References:
HDT: https://corpora.uni-hamburg.de/hzsk/de/islandora/object/treebank:hdt

MrEbbinghaus committed:
Split CoNLL lines at single tabs

CoNLL should be separated by a single tab.
Some CoNLL files (like the Hamburg Dependency Treebank) are treating words like "New York" as a single form with a space!
The CoNLLReader couldn't read those files.

References:
HDT: https://corpora.uni-hamburg.de/hzsk/de/islandora/object/treebank:hdt

dimazest commented:
Looks good. Could you add a test case and add references to the documentation.

alvations commented:
@MrOerni The automated are failing because not all corpora are split on `\t`. It might be better to either allow a `separator` argument when initializing the `ConllCorpusReader` object instead of hardcoding it at https://github.com/MrOerni/nltk/blob/develop/nltk/corpus/reader/conll.py#L197

alvations committed:
Allow customized separator

alvations commented:
[CI: retest]

jnothman commented on nltk/corpus/reader/conll.py:
```diff
@@ -70,7 +70,7 @@ class ConllCorpusReader(CorpusReader):
                  chunk_types=None, root_label='S', pos_in_tree=False,
                  srl_includes_roleset=True, encoding='utf8',
-                 tree_class=Tree, tagset=None):
+                 tree_class=Tree, tagset=None, separator=' '):
```
please use `'\t'` for readability

alvations replied:
The `' '` here is a space =)

alvations replied:
I'll change it to `\s` for readibility.

jnothman reviewed (commented):
It would be good, I think, if this had a non-regression test.

jnothman commented:
But aren't you meant to be splitting on tabs? Have I misunderstood?

jnothman commented:
Oh I see the discussion now. For backwards compatibility you should allow
for regex split.

MrEbbinghaus committed:
Use empty string as default seperator

This maintains the original behavior of split()

MrEbbinghaus commented:
@alvations I changed the default value to `None`. 
With `.split('  ')`, the string is splited by the actual symbol `space`.

With `.split(None)`  the behavior is more like a regex `\s+`.

#### EDIT:
The first version `''` is wrong. `None` should it be. My mistake.

MrEbbinghaus committed:
Change default seperator to None to comply with python3

jnothman commented:
None would be clearer than ''

On 8 January 2018 at 23:40, Björn Ebbinghaus <notifications@github.com>
wrote:

> @alvations <https://github.com/alvations> I changed the default value to
> ''.
> With .split(' '), the string is splited by the actual symbol space.
>
> With .split('') the behavior is more like a regex \s+
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/nltk/nltk/pull/1919#issuecomment-355956217>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz6yNgCK8wWehi6Zf9H3NNDN4huzPaks5tIgy7gaJpZM4RF3Hf>
> .
>

MrEbbinghaus commented:
@jnothman You are right.
`None` is the actual [default value](https://docs.python.org/3/library/stdtypes.html#str.split).

The `''` is actually wrong. I followed https://www.tutorialspoint.com/python/string_split.htm 😞

alvations commented:
Yes `None` is a good default

alvations commented:
[CI: retest]

alvations commented:
[CI: retest]

alvations commented:
@jnothman Thanks for the reviews and the suggestions! 

@MrOerni we had some hiccups on the automated testing. Now it's all cleared up. The tests should pass now =)

MrEbbinghaus commented:
@alvations I was starting to worry... 👍

MrEbbinghaus committed:
Add separator argument to ConllChunkCorpusReader

MrEbbinghaus committed:
Update doc mention of the ConllChunkCorpusReader constructor

MrEbbinghaus committed:
Add documentation for custom separator

alvations commented:
[CI: retest]

alvations commented:
[CI: retest]

alvations commented on nltk/test/corpus.doctest:
```diff
@@ -1472,9 +1472,12 @@ constructor will first call its base class's constructor, and then
+    >>> def __init__(self, root, fileids, chunk_types, encoding='utf8',
+    ...          tagset=None, separator=None):
+    ... ConllCorpusReader.__init__(
+    ...     self, root, fileids, ('words', 'pos', 'chunk'),
```
The indentation is misaligned here.

alvations replied:
Also I think creating a class object in the doctests isn't necessary to test this feature.

alvations committed:
Clean up the doctest

The constructor code doesn't need >>> in the doctest

alvations commented:
The latest changes removed the need for regression test on the constructor in the `corpus.doctest`. 
That should clear the automated test.

alvations commented:
@MrOerni Sorry the automated test took some time to fix. Now it's all cleared up. 

It LGTM now.

stevenbird commented:
Thanks @MrOerni, @alvations, @jnothman, @dimazest
